filament/src/materials/dof.mat

```diff
@@ -49,17 +44,423 @@ fragment {
+        highp vec2 uvCenter, highp vec2 fullResPixelSize, float mip, const bool first) {
+
+    // we accumulate the larger rings first
+    i = (ringCount - 1.0) - i;
+
+    Bucket curr;
+    initBucket(curr);
+
+    float count;
+    vec2 p;
+    mat2 r;
+    initRing(i, ringCount, kernelSize, count, r, p);
+
+    float border = (kernelSize / ringCount) * (i + 0.5);
+    for (float j = 0.0; j < count; j += 2.0) {
+        accumulateBackgroundMirror(curr, prev, uvCenter, p * fullResPixelSize, border, mip, first);
+        p = r * p;
+    }
+
+    mergeRings(curr, prev, count);
+}
+
+/*
+ * Foreground accumulator
+ */
+
+void accumulateForeground(inout vec4 foreground, inout float opacity, inout float i,
+        highp vec2 pos, float border, float mip) {
+    float coc = textureLod(materialParams_cocFgBg, pos, mip).r;
+    vec4 s = textureLod(materialParams_foreground, pos, mip);
+    float inLayer = isForeground(coc);
+    float o = cocToAlpha(coc) * inLayer;
+    float w = intersection(border, abs(coc), mip) * sampleWeight(coc, mip) * inLayer;
+    foreground += s * w;
+    opacity += o;
+    i += w;
+}
+
+void accumulateForegroundCenter(inout vec4 foreground, inout float opacity, inout float i,
+        highp vec2 pos, float border, float mip) {
+    accumulateForeground(foreground, opacity, i,pos, border, mip);
+}
+
+void accumulateForegroundMirror(inout vec4 foreground, inout float opacity, inout float i,
+        highp vec2 center, vec2 offset, float border, float mip) {
+    // The code below is equivalent to:
+    //      accumulateForeground(foreground, opacity, i, center + offset, border, mip);
+    //      accumulateForeground(foreground, opacity, i, center - offset, border, mip);
+    // but selects the min coc of opposite samples as a way to guess the color occluded by
+    // the geometry.
+
+    float coc0 = textureLod(materialParams_cocFgBg, center + offset, mip).r;
+    vec4 s0 = textureLod(materialParams_foreground, center + offset, mip);
+    float coc1 = textureLod(materialParams_cocFgBg, center - offset, mip).r;
+    vec4 s1 = textureLod(materialParams_foreground, center - offset, mip);
+    float coc = min(coc0, coc1);
+    float inLayer = isForeground(coc);
+    float o = cocToAlpha(coc) * inLayer * 2.0;  // we're accumulating 2 samples
+    float w = intersection(border, abs(coc), mip) * sampleWeight(coc, mip) * inLayer;
+    foreground += s0 * w;
+    foreground += s1 * w;
+    opacity += o;
+    i += w * 2.0;
+}
+
+float getMipLevel(float ringCount, float kernelSize) {
+#if KERNEL_USE_MIPMAP
+    float s = 1.0 / (ringCount + 0.5);
+    float mip = log2(kernelSize * s);
+    return mip;
+#else
+    return 0.0;
+#endif
+}
+
 void postProcess(inout PostProcessInputs postProcess) {
+    const float ringCountGather = float(RING_COUNT_GATHER);
+    const float ringCountFast   = float(RING_COUNT_FAST);
     highp vec2 uv = variable_vertex.xy;
+    vec2 tiles = textureLod(materialParams_tiles, uv, 0.0).rg;
```

Comment:
Nit: I would have called this variable `tileInfo` rather than `tiles` to better indicate that it holds CoC information for a single tile.